jax/lax/lax_parallel.py

```diff
@@ -579,17 +579,20 @@ def _moveaxis(src, dst, x):
 pxla.multi_host_supported_collectives.add(all_to_all_p)
 
 
-def _expand(dim, size, axis_name, x):
+def _expand(dim, size, index, x):
   shape = list(x.shape)
   shape.insert(dim, size)
   out = lax.full(shape, lax._const(x, 0))
-  return lax.dynamic_update_index_in_dim(out, x, axis_index(axis_name), dim)
+  return lax.dynamic_update_index_in_dim(out, x, index, dim)
 
-def _allgather(x, dim, size, axis_name):
-  outs = tree_util.tree_map(partial(_expand, dim, size, axis_name), x)
-  return psum(outs, axis_name)
+def _allgather(x, dim, size, axis_name, axis_index_groups=None):
+  index = axis_index(axis_name)
+  if axis_index_groups is not None:
+    index = index % size
```

Comment:
This area is new to me, but I am wondering if this clamping computation is correct. I assume that this is needed because `axis_index` is agnostic of the `axis_index_groups`. Then I suspect that the clamping computation is not correct. E.g., if `axis_index_groups=[[0, 2], [1, 3]`. Am I misunderstanding something?

Reply:
The reason is indeed because `axis_index` doesn't account for `axis_index_groups`.
As a workaround I've added a separate `axis_index_to_group_index` kwarg which provides the necessary mapping from global to group index. I've also added a test case that covers your example.